python/mleap/pyspark/feature/string_map.py

```diff
@@ -1,17 +1,36 @@
+import six
 from pyspark.ml.util import JavaMLReadable, JavaMLWritable, _jvm
 from pyspark.ml.wrapper import JavaTransformer
 from pyspark.ml.param.shared import HasInputCol, HasOutputCol
+from pyspark.sql import DataFrame
 
 
 class StringMap(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):
-    def __init__(self, labels, inputCol=None, outputCol=None, handleInvalid='error', defaultValue=0.0):
+    def __init__(self, labels={}, inputCol=None, outputCol=None, handleInvalid='error', defaultValue=0.0):
```

Comment:
I'm not sure that all of these parameters are preserved after serialization, have you tested if `handleInvalid` and `defaultValue` are the same after deserializing ?

I don't see how the python `StringMap` can be initialized correctly after deserialization because those 2 fields are not serialized in the first place. They are not part of the StringMap JavaTransformer here: https://github.com/combust/mleap/blob/d276c1c800277a1edcf30dda6e8559c7413b21f8/mleap-spark-extension/src/main/scala/org/apache/spark/ml/mleap/feature/StringMap.scala so I think the just get lost and the deserialized object will always have the defaults `error` and 0.0

(I also see the deserialization test is inactive, see https://github.com/combust/mleap/pull/624)

Reply:
@lucagiovagnoli thanks for looking into this.

> I'm not sure that all of these parameters are preserved after serialization, have you tested if handleInvalid and defaultValue are the same after deserializing ?

Good question. I think I tested it only in real life and forgot to add such case in the unit test. I have added it to the test now in the open PR: https://github.com/combust/mleap/pull/624/commits/49cd0139df7a41f31b81d5b3b91f3a3526e78659

> I don't see how the python StringMap can be initialized correctly... https://github.com/combust/mleap/blob/d276c1c800277a1edcf30dda6e8559c7413b21f8/mleap-spark-extension/src/main/scala/org/apache/spark/ml/mleap/feature/StringMap.scala

For some reason that link is to an old commit? Note that handleInvalid and defaultValue are part of StringMapModel, see https://github.com/combust/mleap/blob/master/mleap-core/src/main/scala/ml/combust/mleap/core/feature/StringMapModel.scala.

(if you check StringMapModel in the same commit as your link, they are not there yet https://github.com/combust/mleap/blob/d276c1c800277a1edcf30dda6e8559c7413b21f8/mleap-core/src/main/scala/ml/combust/mleap/core/feature/StringMapModel.scala)

> I also see the deserialization test is inactive

Oh, but the comment already said that it works and is just about test setup in Travis. It was initially disabled for Travis because of a problem related to multi-scala build. You can see that the Travis build has succeeded for commit 24a6ad5 in https://github.com/combust/mleap/pull/624, which has a fix for the class path setup. The last commit was just updating with master, but somehow Travis failed to start a build entirely, meh. But as I said you can see the proof that test_serialize_to_bundle works in that PR. It tests that 'a' is mapped to 1.0 by the deserialized StringMap.

Now with my new commit Travis build seems to have started. I hope there are no random infrastructure failures this time and it passes.

I haven't heard about @ancasarb in a long time. I wonder how to get that PR merged.

Reply:
I think you might be right, in that previous blob, loading 'handleInvalid' here https://github.com/combust/mleap/blob/6d2725aa9dff9bbb0f7deec465f49edf0ec78107/mleap-runtime/src/main/scala/ml/combust/mleap/bundle/ops/feature/StringMapOp.scala#L33 was missing but I see you've added it